src/v/raft/log_eviction_stm.cc

```diff
@@ -16,23 +16,161 @@
+    return raft::state_machine::start();
+}
+
+static bool is_within_bounds(
+  ss::logger& logger,
+  ss::lw_shared_ptr<const storage::offset_translator_state> translator,
+  model::offset kafka_truncation_offset,
+  model::offset start_offset,
+  model::offset high_watermark) {
+    const auto kafka_start_offset = translator->from_log_offset(start_offset);
+    const auto kafka_high_offset = translator->from_log_offset(high_watermark);
+    vlog(
+      logger.trace,
+      "Truncation offset: {} start offset: {} high watermark: {} kafka start "
+      "offset: {} kafka high watermark: {}",
+      kafka_truncation_offset,
+      start_offset,
+      high_watermark,
+      kafka_start_offset,
+      kafka_high_offset);
+    return kafka_truncation_offset > kafka_start_offset
+           && kafka_truncation_offset <= kafka_high_offset;
+}
+
+ss::future<std::error_code> log_eviction_stm::truncate(
+  model::offset kafka_truncate_offset,
+  ss::lowres_clock::time_point deadline,
+  std::optional<std::reference_wrapper<ss::abort_source>> as) {
+    /// The offset provided by the user is the new start offset, everything
+    /// before it should be truncated
+    if (!is_within_bounds(
+          _logger,
+          _raft->get_offset_translator_state(),
+          kafka_truncate_offset,
+          _raft->start_offset(),
+          model::next_offset(_raft->last_visible_index()))) {
+        vlog(
+          _logger.info,
+          "Cannot prefix_truncate at offset: {}",
+          kafka_truncate_offset);
+        co_return make_error_code(raft::errc::invalid_truncation_offset);
+    }
+    /// Create the special prefix_truncate batch, it is a model::record_batch
+    /// with exactly one record within it, the point at which to truncate
+    storage::record_batch_builder builder(
+      model::record_batch_type::prefix_truncate, model::offset(0));
+    /// Everthing below the requested offset should be truncated, requested
+    /// offset itself will be the new low_watermark (readable)
+    const model::offset as_log_offset
+      = _raft->get_offset_translator_state()->to_log_offset(
+        kafka_truncate_offset);
+    auto value = serde::to_iobuf(as_log_offset - model::offset{1});
+    builder.add_raw_kv(iobuf(), std::move(value));
+    auto batch = std::move(builder).build();
+    /// After command replication all that can be guaranteed is that the command
+    /// was replicated
+    auto result = co_await replicate_commands(std::move(batch), deadline, as);
+    if (result) {
+        co_return result;
+    }
+    try {
+        /// The following will return when the logs start_offset has
+        /// officially been incremented at or above the requested truncated
+        /// offset on this node. No guarantees of data removal / availability
+        /// can be made at or after this point.
+        co_await _raft->min_visible_offset_monitor().wait(
+          as_log_offset, deadline, as);
```

Comment:
it looks like we do not need the `min_visible_offset_monitor` the wait in `replicate_commands` should be enough.

Reply:
I did this because the `wait` in `replicate_commands` will return when the log end moves forward. This portion of code is waiting on the start offset to move forward. The start offset is already far behind the head so if I do this logic with `wait`, the command immediately returns and this will return to the client before the effect was actually performed.